vscode/src/completions/providers/unstable-azure-openai.ts

```diff
@@ -0,0 +1,107 @@
+import { Completion } from '..'
+import { logger } from '../../log'
+import { ReferenceSnippet } from '../context'
+import { getHeadAndTail } from '../text-processing'
+
+import { Provider, ProviderConfig, ProviderOptions } from './provider'
+
+interface UnstableAzureOpenAIOptions {
+    serverEndpoint: string
+    accessToken: string
+}
+
+interface AzureOpenAIApiResponse {
+    id: string
+    object: 'text_completion'
+    created: number
+    model: string
+    choices: {
+        text: string
+        index: number
+        finish_reason: 'stop'
+        logprobs: null
+    }[]
+    usage: {
+        completion_tokens: number
+        prompt_tokens: number
+        total_tokens: number
+    }
+}
+
+const PROVIDER_IDENTIFIER = 'azure-openai'
+
+export class UnstableAzureOpenAIProvider extends Provider {
+    private serverEndpoint: string
+    private accessToken: string
+
+    constructor(options: ProviderOptions, unstableAzureOpenAIOptions: UnstableAzureOpenAIOptions) {
+        super(options)
+        this.serverEndpoint = unstableAzureOpenAIOptions.serverEndpoint
+        this.accessToken = unstableAzureOpenAIOptions.accessToken
+    }
+
+    public async generateCompletions(abortSignal: AbortSignal, snippets: ReferenceSnippet[]): Promise<Completion[]> {
+        const { head, tail } = getHeadAndTail(this.options.prefix)
+
+        // Create prompt
+        const prompt = `Human: You are a senior engineer assistant working on a codebase. Complete the following code:\n\n${head.trimmed}\n\nAssistant:\n${tail.trimmed}`
+
+        // Issue request
+        const request = {
+            prompt,
+            temperature: 1,
+            top_p: 0.5,
+            frequency_penalty: 0,
+            presence_penalty: 0,
+            max_tokens: 100,
```

Comment:
We probably need more for multi-line requests and can do with less for single-line

```suggestion
            max_tokens: this.config.multiline ? 256 : 50,
```